### PR TITLE
Fix #7146: support (...) in type hint comments (V2)

### LINKED ISSUE #7146: autodoc.type_comments throw IndexError
**Describe the bug**
Since 2.4.0, we get a failure on our test suites:
```
Warning, treated as error:
1116error while formatting arguments for scapy.contrib.http2.HPackHdrTable.parse_txt_hdrs: list index out of range
```
I mocked sphinx to display the stack trace:
```
Exception occurred:
2914  File "/home/travis/build/secdev/scapy/.tox/docs/lib/python3.8/site-packages/Sphinx-2.4.2.dev20200213-py3.8.egg/sphinx/ext/autodoc/type_comment.py", line 62, in update_annotations_using_type_comments
2915    annotation = ast_unparse(function.argtypes[i])  # type: ignore
```

Here's what the function looks like:
https://github.com/secdev/scapy/blob/0dd974e46b3678f02e422dacafc507f32cacde89/scapy/contrib/http2.py#L2636

The issue happens whenever there is a mismatch between the amount of arguments and the amount of type hints comments. This happens when using the `(...)`

This triggers it:
```python

def pony(a,
         b,  # type: str
         c=None  # type: Optional[str]
         ):
    # type: (...) -> str
    return b + (c or "")
```

This is an issue when using functions within class. You don't put a type hint for `self` (https://mypy.readthedocs.io/en/stable/cheat_sheet.html#classes)

<details>

<summary>Full travis logs</summary>

https://travis-ci.com/secdev/scapy/jobs/286619573
Backup:
https://pastebin.com/raw/vxhT1cGq

</details>

**To Reproduce**

```
$ git clone https://github.com/gpotter2/issues-sphinx.git
$ cd issues-sphinx
$ make html
```

**Your project**
https://github.com/secdev/scapy

**Environment info**
- OS: linux
- Python version: 3.8.1
- Sphinx version: 2.4.1
- Sphinx extensions:  sphinx.ext.autodoc

Thanks for the time.

### PR DESCRIPTION
**Pull request by gpotter2**
An alternative fix to #7146, with tests.

I prefer this one: there is no compromises like the ones in #7147 (we don't re build then re parse the arguments, and we never fail to bind the correct attribute to the correct name).

However this changes the output of `get_type_comment`. I haven't found other uses of the function.
It could also return a tuple instead of the dictionary. I'm open to changes

### CODE REVIEW

File: sphinx/ext/autodoc/type_comment.py [hunk 1]
```diff
@@ -41,7 +42,34 @@ def get_type_comment(obj: Any) -> ast.FunctionDef:
...
+                node = subject
+                results = {
+                    'returns': node.returns,
+                    'explicit': True,
+                    'args': {
+                        arg.arg: arg.type_comment
+                        for arg in itertools.chain(
+                            getattr(node.args, "posonlyargs", []) or [],
+                            getattr(node.args, "args", []) or [],
+                            getattr(node.args, "vararg", []) or [],
```
Comment (tk0miya) [inline sphinx/ext/autodoc/type_comment.py]: vararg and kwarg are not list. So this code would not work. Please test with `(*args, **kwargs)`.

File: sphinx/ext/autodoc/type_comment.py [hunk 2]
```diff
@@ -41,7 +42,34 @@ def get_type_comment(obj: Any) -> ast.FunctionDef:
...
+            else:
+                node = subject
+                results = {
+                    'returns': node.returns,
+                    'explicit': True,
+                    'args': {
+                        arg.arg: arg.type_comment
+                        for arg in itertools.chain(
+                            getattr(node.args, "posonlyargs", []) or [],
+                            getattr(node.args, "args", []) or [],
```
Comment (tk0miya) [inline sphinx/ext/autodoc/type_comment.py]: Except posonlyargs, other attributes (ex. args, varargs) always exist. So you don't need to call `getattr()`. It makes code difficult to read (a bit).

Note: posonlyargs exists only python3.8 and later.

File: sphinx/ext/autodoc/type_comment.py [hunk 3]
```diff
@@ -41,7 +42,34 @@ def get_type_comment(obj: Any) -> ast.FunctionDef:
...
+                    ast.FunctionDef,
+                    ast_parse(comment, mode='func_type')
+                )
+                results = {
+                    'returns': node.returns,
+                    'explicit': False,
+                    'args': node.argtypes  # type: ignore
+                }
+            else:
+                node = subject
```
Comment (tk0miya) [inline sphinx/ext/autodoc/type_comment.py]: Why did you rename `subject` here?

File: sphinx/ext/autodoc/type_comment.py [hunk 4]
```diff
@@ -41,7 +42,34 @@ def get_type_comment(obj: Any) -> ast.FunctionDef:
             subject = cast(ast.FunctionDef, module.body[0])  # type: ignore
 
         if getattr(subject, "type_comment", None):
-            return ast_parse(subject.type_comment, mode='func_type')  # type: ignore
+            comment = subject.type_comment  # type: ignore
```
Comment (tk0miya) [inline sphinx/ext/autodoc/type_comment.py]: Why did you make a new variable `comment` here? It is used only twice (in first if-block).

File: tests/test_ext_autodoc_configs.py [hunk 5]
```diff
@@ -498,6 +506,10 @@ def test_autodoc_typehints_signature(app):
         '',
         '.. py:function:: incr(a: int, b: int = 1) -> int',
         '   :module: target.typehints',
+        '',
+        '',
+        '.. py:function:: missing_attr(c: None, a: str, b: Optional[str] = None) -> None',
```
Comment (tk0miya) [inline tests/test_ext_autodoc_configs.py]: I feel strange for non-annotated arguments are typed as `None`. I think they should not be annotated.

Comment (tk0miya) [inline tests/test_ext_autodoc_configs.py]: Hmm? It seems the type_comment of `missing_attr()` is `# type: (...) -> str`. The returns annotation should be `str`. Is this testcase correct?

**Comment by tk0miya** [comment]
As a trial, I just made a patch to improve this: 98d2446. Please merge it if you like it.

**Comment by gpotter2** [comment]
That's a much cleaner patch. Great job, thanks a lot

**Comment by tk0miya** [comment]
Thank you for update. LGTM!! I'll release this as 2.4.2 within a few days :-)

**Comment by gpotter2** [comment]
Great thanks !
